{% extends "layout.html" %}
{% block head %}

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.0/css/bootstrap.min.css" integrity="sha512-gSfrSyGfjdjA1G2c+eP7498HCNt+5sJ23zTW7qr2zuO2P7ntIcSGOXP2rHXRY3jh3o36jYRWs8mYX8LEVrAkFA==" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.tablesorter/2.31.3/js/jquery.tablesorter.min.js" integrity="sha512-qzgd5cYSZcosqpzpn7zF2ZId8f/8CHmFKZ8j7mU4OUXTNRd5g+ZHBPsgKEwoqxCtdQvExE5LprwwPAgoicguNg==" crossorigin="anonymous"></script>

<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.tablesorter/2.31.3/js/jquery.tablesorter.widgets.min.js" integrity="sha512-dj/9K5GRIEZu+Igm9tC16XPOTz0RdPk9FGxfZxShWf65JJNU2TjbElGjuOo3EhwAJRPhJxwEJ5b+/Ouo+VqZdQ==" crossorigin="anonymous"></script>

<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.tablesorter/2.31.3/js/extras/jquery.tablesorter.pager.min.js" integrity="sha512-y845ijdup9lDunrcSRQAlFdQICHVhkB5UNguWRX8A3L+guxO7Oow0poojw0PLckhcKij++h85bnyro80fjR9+A==" crossorigin="anonymous"></script>

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/jquery.tablesorter/2.31.3/css/theme.bootstrap_3.min.css" integrity="sha512-1r2gsUynzocV5QbYgEwbcNGYQeQ4jgHUNZLl+PMr6o248376S3f9k8zmXvsKkU06wH0MrmQacKd0BjJ/kWeeng==" crossorigin="anonymous" />

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/jquery.tablesorter/2.31.3/css/jquery.tablesorter.pager.min.css" integrity="sha512-TWYBryfpFn3IugX13ZCIYHNK3/2sZk3dyXMKp3chZL+0wRuwFr1hDqZR9Qd5SONzn+Lja10hercP2Xjuzz5O3g==" crossorigin="anonymous" />

<script>
  $(function() {

    // NOTE: $.tablesorter.themes.bootstrap is ALREADY INCLUDED in the jquery.tablesorter.widgets.js
    // file; it is included here to show how you can modify the default classes
    $.tablesorter.themes.bootstrap = {
      // these classes are added to the table. To see other table classes available,
      // look here: http://getbootstrap.com/css/#tables
      table        : 'table table-bordered table-striped',
      caption      : 'caption',
      // header class names
      header       : 'bootstrap-header', // give the header a gradient background (theme.bootstrap_2.css)
      sortNone     : '',
      sortAsc      : '',
      sortDesc     : '',
      active       : '', // applied when column is sorted
      hover        : '', // custom css required - a defined bootstrap style may not override other classes
      // icon class names
      icons        : '', // add "bootstrap-icon-white" to make them white; this icon class is added to the <i> in the header
      iconSortNone : 'bootstrap-icon-unsorted', // class name added to icon when column is not sorted
      iconSortAsc  : 'glyphicon glyphicon-chevron-up', // class name added to icon when column has ascending sort
      iconSortDesc : 'glyphicon glyphicon-chevron-down', // class name added to icon when column has descending sort
      filterRow    : '', // filter row class; use widgetOptions.filter_cssFilter for the input/select element
      footerRow    : '',
      footerCells  : '',
      even         : '', // even row zebra striping
      odd          : ''  // odd row zebra striping
    };

    // call the tablesorter plugin and apply the uitheme widget
    $("#table_prs_trait_list").tablesorter({
      // this will apply the bootstrap theme if "uitheme" widget is included
      // the widgetOptions.uitheme is no longer required to be set
      theme : "bootstrap",

      widthFixed: true,

      headerTemplate : '{content} {icon}', // new in v2.7. Needed to add the bootstrap icon!

      // widget code contained in the jquery.tablesorter.widgets.js file
      // use the zebra stripe widget if you plan on hiding any rows (filter widget)
      widgets : [ "uitheme", "filter", "columns", "zebra" ],

      widgetOptions : {
        // using the default zebra striping class name, so it actually isn't included in the theme variable above
        // this is ONLY needed for bootstrap theming if you are using the filter widget, because rows are hidden
        zebra : ["even", "odd"],

        // class names added to columns when sorted
        columns: [ "primary", "secondary", "tertiary" ],

        // reset filters button
        filter_reset : ".reset",

        // extra css class name (string or array) added to the filter element (input or select)
        filter_cssFilter: "form-control",

        // set the uitheme widget to use the bootstrap theme class names
        // this is no longer required, if theme is set
        // ,uitheme : "bootstrap"

      }
    })
    .tablesorterPager({

      // target the pager markup - see the HTML block below
      container: $(".ts-pager"),

      // target the pager page select dropdown - choose a page
      cssGoto  : ".pagenum",

      // remove rows from the table to speed up the sort of large tables.
      // setting this to false, only hides the non-visible rows; needed if you plan to add/remove rows with the pager enabled.
      removeRows: false,

      // output string - default is '{page}/{totalPages}';
      // possible variables: {page}, {totalPages}, {filteredPages}, {startRow}, {endRow}, {filteredRows} and {totalRows}
      output: '{startRow} - {endRow} / {filteredRows} filtered traits  (out of {totalRows} available traits)'

    });
  });

  </script>
{% endblock %}

{% block body %}

<div class="container">
  <div class="row">
    <div class="col col-md-12">
      <h1>Sparse Polygenic Risk Scores (PRSs) with BASIL / snpnet</h1>
      <hr/>
      <h3>Version 2.1.0, 2022.01.19</h3>
      <h4>
        If you have used the previous version of the results, please check <a href="/prs_v1">our browser for version 1.3.1</a> and the description of the changes in <a href="#version_info">version info section</a> below.
      </h4>
      <p>Yosuke Tanigawa and Manuel A. Rivas</p>
      <hr/>
    </div>
  </div>

  <div class="row">
    <div class="col col-md-12">
      <h2>Data download</h2>
      <p>
        You can download the coefficient of the PRS model using the "download" button on each of the PRS model page.
      </p>
      <p>
        The coefficients of the 428 significant PRS models (from version 1 of the analysis) are available at <a href="https://www.pgscatalog.org/publication/PGP000244/" target="_blank">PGS catalog (PGS catalog score IDs: PGS000922-PGS001349)</a>
      </p>
      <hr/>
    </div>
  </div>

  <div class="row">
    <div class="col col-md-12">
      <h2 align="center">Browseable phenotypes</h2>
      <p>
        Here, we display available sparse PRS models in UK Biobank.
        One can use sort and/or filter function. For example, you may enter ">1000" in the '# variants' column to select PRS models with more than 1,000 genetic variants.
      </p>
      <br>
      <div id="container_prs_summary" style="font-size:100%;">
        <!-- https://mottie.github.io/tablesorter/docs/index.html -->
        <table id="table_prs_trait_list"> <!-- bootstrap classes added by the uitheme widget -->
          <thead>
            <tr>
              <th colspan="2">Trait</th>
              <th rowspan="2" class="filter-select filter-exact" data-placeholder="Select">Family</th>
              <th colspan="4">Predictive performance</th>
              <th rowspan="2"># variants</th>
              <th colspan="2">Significance</th>
            </tr>
            <tr>
              <th class="filter-select filter-exact" data-placeholder="Select">Trait group</th>
              <th>Trait name</th>
              <th>Geno</th>
              <th>Covars</th>
              <th>Full</th>
              <th>delta</th>
              <th>p (WB)</th>
              <th class="filter-select filter-exact" data-placeholder="Select">significant?</th>
            </tr>
          </thead>
          <tfoot>
            <tr>
              {% for table_col in table_prs_trait_list_cols %}
              <th>{{ table_col }}</th>
              {% endfor %}
            </tr>
            <tr>
              <th colspan="{{ table_prs_trait_list_col_len }}" class="ts-pager form-inline">
                <div class="btn-group btn-group-sm" role="group">
                  <button type="button" class="btn btn-default first"><< First</button>
                  <button type="button" class="btn btn-default prev">< Prev.</button>
                </div>
                <span class="pagedisplay"></span>
                <div class="btn-group btn-group-sm" role="group">
                  <button type="button" class="btn btn-default next">Next ></button>
                  <button type="button" class="btn btn-default last">Last >></button>
                </div>
                <select class="form-control input-sm pagesize" title="Select page size">
                  <option selected="selected" value="10">10</option>
                  <option value="30">30</option>
                  <option value="50">50</option>
                  <option value="100">100</option>
                  <option value="all">All Rows</option>
                </select>
                <select class="form-control input-sm pagenum" title="Select page number">Ã¥
                </select>
              </th>
            </tr>
          </tfoot>
          <tbody>
            {{ table_prs_trait_list_tbody_str|safe }}
          </tbody>
        </table>
      </div>
    </div>
  </div>
  <div class="row">
    <div class="col col-md-12">
      <h3 align="center">Description of the columns</h3>
      <ul>
          <li>
            Trait group: trait categories used in the Global Biobank Engine. Its <a href="https://gbe.stanford.edu/faq#group" target="_blank">FAQ page</a> has more information.
          </li>
          <li>
            Trait: a trait in UK Biobank.
          </li>
          <li>
            Family: it represents whether the PRS model is fit with gaussian model (for continuous response, such as standing height) or binomial model (for binary outcomes, such as disease). We use <i>R</i><sup>2</sup> and Nagelkerke's pseudo-<i>R</i><sup>2</sup> (also known as Cragg and Uhler's pseudo-<i>R</i><sup>2</sup>) as the evaluation metric for gaussian and binomial models, respectively. In this table, we display those evaluation metric computed on the hold-out test set consits of individuals of whiter British ancestry in UK Biobank cohort.
          </li>
          <li>
            Geno: the predictive performence (<i>R</i><sup>2</sup> or Nagelkerke's pseudo-<i>R</i><sup>2</sup>) of the predictive model with genotype features alone. Note that this model does not include genotype PCs.
          </li>
          <li>
            Covar: the predictive performence (<i>R</i><sup>2</sup> or Nagelkerke's pseudo-<i>R</i><sup>2</sup>) of the predictive model with covariates (age, sex, and genotype PCs) alone.
          </li>
          <li>
            Full: the predictive performence (<i>R</i><sup>2</sup> or Nagelkerke's pseudo-<i>R</i><sup>2</sup>) of the full model that have both genotypes and covariates alone.
          </li>
          <li>
            delta: the incremental predictive performence (<i>R</i><sup>2</sup> or Nagelkerke's pseudo-<i>R</i><sup>2</sup>) of the full model against the covariate-only model, i.e. the difference between the values in "Full" and "Covar" columns.
          </li>
          <li>
            # variants: the number of genetic variants included in the model
          </li>
          <li>
            P (WB): P-value of the incremental predictive performance of the PRS model evaluated on the test set individuals of white British ancestry in UK Biobank. The p-value is before the multiple hypothesis correction.
          </li>
          <li>
            significant?: this indicator variable shows whether the PRS model has a significant (P < 0.05/2000) incremental predictive performance.
          </li>
      </ul>
      <br>
      <hr/>
    </div>
  </div>

  <div class="row">
    <div class="col col-md-12">
      <h2>
        Comparison of the estimated SNP-based heritability and the the predictive performance of the sparse PRS
      </h2>
    </div>
  </div>
  <div class="row">
    <div class="col col-md-6">
      <a href="http://web.stanford.edu/~ytanigaw/PRSmap/PRSmap_v2/h2_vs_geno_NagelkerkeR2.html" target="_blank">
        <img class="img-fluid" width="100%" src="/static/PRSmap/PRSmap_v2/h2_vs_geno_NagelkerkeR2.png" alt="binary traits"></img>
      </a>
    </div>
    <div class="col col-md-6">
      <a href="http://web.stanford.edu/~ytanigaw/PRSmap/PRSmap_v2/h2_vs_geno_gaussian.html" target="_blank">
        <img class="img-fluid" width="100%" src="/static/PRSmap/PRSmap_v2/h2_vs_geno_gaussian.png" alt="quantitative traits"></img>
      </a>
    </div>
  </div>

  <div class="row">
    <div class="col col-md-12">
      <p>
        Comparison of the estimated SNP-based heritability and predictive performance across the 813 traits with significant PRSs. The predictive performance (Nagelkerke's pseudo-<i>R</i><sup>2</sup> for 244 binary traits [left] and <i>R</i><sup>2</sup> for 569 quantitative traits [right]) of the PRS models that only consider genetic variants are compared against the estimated SNP-based heritability. The gray solid lines represent y = x line. The points on the bottom left corners were shown in the inset plots. The error bars represent standard error. BMD: Bone mineral density.
      </p>
      <hr/>
    </div>
  </div>

  <div class="row">
    <div class="col col-md-12">
      <h2>Comparison of the effect size and the model size of the significant sparse PRS</h2>
    </div>
  </div>
  <div class="row">
    <div class="col col-md-6">
      <a href="http://web.stanford.edu/~ytanigaw/PRSmap/PRSmap_v2/size_vs_delta_NagelkerkeR2.html" target="_blank">
        <img class="img-fluid" width="100%" src="/static/PRSmap/PRSmap_v2/size_vs_delta_NagelkerkeR2.png" alt="binary traits"></img>
      </a>
    </div>
    <div class="col col-md-6">
      <a href="http://web.stanford.edu/~ytanigaw/PRSmap/PRSmap_v2/size_vs_delta_gaussian.html" target="_blank">
        <img class="img-fluid" width="100%" src="/static/PRSmap/PRSmap_v2/size_vs_delta_gaussian.png" alt="quantitative traits"></img>
      </a>
    </div>
  </div>

  <div class="row">
    <div class="col col-md-12">
      <p>
        Comparison of the effect size and the model size of sparse PRS. The number of the genetic variants included in the model (size of the model, x-axis) and the incremental predictive performance (effect size of the model, y-axis) are shown for 244 binary traits (left) and 569 quantitative traits (right). TTE: time-to-event phenotype.
      </p>
      <hr/>
    </div>
  </div>

  <div class="row">
    <div class="col col-md-12">
      <a name="version_info"></a>
      <h2>Version info</h2>
      <ul>
        <li>
          2022.01.19. Version 2.1.0: We changed the main predictive performance metric for binary traits to Nagelkerke's pseudo-<i>R</i><sup>2</sup> from ROC-AUC. We made minor stylistic updates in the figures.
        </li>
        <li>
          2021.11.04. Version 2.0.0: We made several changes. Of note, the coefficients of the PRS models does not change for most traits with the exception of the biomarker traits (please read (3) below). (1) We found a bug in our performance evaluation script, which resulted in inaccurate statistical significance reported in the previous version. After fixing the issue, we saw an increase in significance for many PRS models, resulting in total of 813 significant (p < 2.5e-5) PRS models. (2) Starting this version, we have also included an indicator variable denoting the types of genotyping arrays in covariates when evaluating the predictive performance. (3) For the blood and urine biomarker traits analyzed in <a href="https://doi.org/10.1038/s41588-020-00757-z">our previous study (Sinnott-Armstrong*, Tanigawa*, et al 2021)</a>, we used two versions of traits (one from the raw measurement values and the other after the covariate adjustment). Starting this version, we have removed those duplicates. Specifically, we took the PRS models from our biomarker paper, which we computed for the covariate-adjusted traits, and evaluated their performance for the raw phenotypes, so that the incremental predictive performance has meaningful values.
        </li>
        <li>
          2021.09.23. Version 1.3.1: We improved the clarity of the table and Website. We added a link to our manuscript.
        </li>
        <li>
          2021.08.23. Version 1.3: We add plots that compares the number of active variables in the PRS model vs the incremental predictive performance of the PRS model.
        </li>
        <li>
          2021.07.23. Version 1.2: We add the p-value of the PRS model as Well as an indicator variable of the significance of PRS. We also removed some of the duplicated traits.
        </li>
        <li>
          2021.01.16. Version 1.1: Initial public release of the results
        </li>
      </ul>
      <hr/>
    </div>
  </div>

  <div class="row">
    <div class="col col-md-12">
      <h2>References</h2>
      <ul>
        <li>
          Y. Tanigawa, J. Qian, G. R. Venkataraman, J. M. Justesen, R. Li, R. Tibshirani, T. Hastie, M. A. Rivas, Significant Sparse Polygenic Risk Scores across 428 traits in UK Biobank. medRxiv 2021.09.02.21262942 (2021). <a href="https://doi.org/10.1101/2021.09.02.21262942" target="_blank">doi:10.1101/2021.09.02.21262942</a>
          <ul>
            <li>
              This manuscript describes the phenome-wide application of BASIL algorithm implemented in the R snpnet package. Please consider citing this paper if you use the PRS models listed on this website.
            </li>
          </ul>
        </li>
        <li>
          J. Qian, Y. Tanigawa, W. Du, M. Aguirre, C. Chang, R. Tibshirani, M. A. Rivas, T. Hastie, A fast and scalable framework for large-scale and ultrahigh-dimensional sparse regression with application to the UK Biobank. PLOS Genetics. 16, e1009141 (2020). <a href="https://doi.org/10.1371/journal.pgen.1009141" target="_blank">doi:10.1371/journal.pgen.1009141</a>
          <ul>
            <li>
              This paper describes the BASIL algorithm and its implementation in the R snpnet package.
            </li>
          </ul>
        </li>
        <li>
          N. Sinnott-Armstrong, Y. Tanigawa, D. Amar, N. J. Mars, C. Benner, M. Aguirre, G. R. Venkataraman, M. Wainberg, H. M. Ollila, T. Kiiskinen, A. S. Havulinna, J. P. Pirruccello, J. Qian, A. Shcherbina, FinnGen, F. Rodriguez, T. L. Assimes, V. Agarwala, R. Tibshirani, T. Hastie, S. Ripatti, J. K. Pritchard, M. J. Daly, M. A. Rivas, Genetics of 35 blood and urine biomarkers in the UK Biobank. Nature Genetics. 53(2), 185-194 (2021). <a href="https://doi.org/10.1038/s41588-020-00757-z" target="_blank">doi:10.1038/s41588-020-00757-z</a> (<a href="https://rdcu.be/cdMmM" target="_blank">full text via ReadCube</a>)
          <ul>
            <li>
              This paper describes the PRS models for the 35 biomarker traits. The definition of the ancestry groups used in the transferability analysis is also described in this paper.
            </li>
          </ul>
        </li>
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col col-md-12">
      <!-- <h2>Methods summary</h2>
      <h3>Construction of the sparse PRS models</h3>
      <h3>Evaluation of the PRS models</h3> -->
    </div>
  </div>
</div>
{% endblock %}

