{% extends "layout.html" %}
{% block head %}
    <!-- Render context vars in JS here -->
    <script type="text/javascript">
        window.icd = {{ icd|tojson|safe }};
        window.page_name = window.icd[0].icd_id;
    </script>
    <script type="text/javascript" src="https://cdn.plot.ly/plotly-latest.min.js"></script>
    <script>
        filtered_variants = _.filter(window.table_variants, function(d) {
            return d.filter == 'PASS';
        });
        $('#number_of_variants').replaceWith(filtered_variants.length);
        lof_filtered_variants = _.filter(window.table_variants, function(d) {
            return d.filter == 'PASS' && d.category == 'lof_variant';
        });$('#number_of_lof_variants').replaceWith(lof_filtered_variants.length);
    </script>

    <style>
        table {
            border-collapse: collapse;
            font: sans-serif;
        }
        th {
            border: 1px black solid;
            padding: 5px;
            text-align: center;
        }
        td {
            border: 1px black solid;
            padding: 5px;
        }

        @font-face {
            font-family: 'Glyphicons Halflings';
            src: url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.eot');
            src: url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.eot?#iefix') format('embedded-opentype'),
                url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.woff2') format('woff2'),
                url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.woff') format('woff'),
                url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.ttf') format('truetype'),
                url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.svg#glyphicons_halflingsregular') format('svg');
        }

        .glyphicon {
            position: relative;
            top: 1px;
            display: inline-block;
            font: normal normal 16px/1 'Glyphicons Halflings';
            -moz-osx-font-smoothing: grayscale;
            -webkit-font-smoothing: antialiased;
            margin-right: 4px;
        }
        /* Add icons you will be using below */
        .glyphicon-fire:before {
            content: '\e104';
        }
        .glyphicon-eye-open:before {
            content: '\e105';
        }
        .glyphicon-cloud-download:before {
            content: '\e197';
        }
    </style>
{% endblock %}

{% block body %}
<div class="container">
    <div class="row">
        <div class="col col-md-12">
            <h1>Sparse Polygenic Risk Scores (PRSs) with BASIL / snpnet</h1>
            <hr/>
            <h3>Version 2.0.0, 2021.11.04</h3>
            <h4>
                If you have used the previous version of the results, please check <a href="/prs_v1">our browser for version 1.3.1</a>
            </h4>
            <p>Yosuke Tanigawa and Manuel A. Rivas</p>
        </div>
    </div>

    <div class="row">
        <div class="col col-md-12">
            <h1>Phenotype: {{icd[0].Name}}</h1>
            <p><a href="/prs">Go back to the PRS top page</a></p>
            <hr/>
        </div>
    </div>

    <div class="row">
        <div class="col col-md-6">
            <dl class="dl-horizontal" style="margin-bottom: 0px;">
                <dt>Phenotype code:</dt>
                <dt>Case Count or N:</dt>
                <dt>Single variant:</dt>
                <dd>{{ icd[0].icd }}</dd>
                <dd>{{ icd[0].Case }}</dd>
                <dd><a href="/{{namespace}}/coding/{{ icd[0].icd}}">{{ icd[0].icd.replace('ICD','') }}</a></dd>
            </dl>
        </div>
        <div class="col col-md-6">
            <a href="{{ snpnet_BETAs_tsv }}" class="btn btn-block btn-primary">
                <span class="glyphicon glyphicon-cloud-download"></span> {{icd[0].Name}} PRS weights
            </a>
        </div>
    </div>

    <div class="row">
        <div class="col col-md-12">
            <hr/>
        </div>
    </div>

    <div class="row">
        <div class="col col-md-12">
            <h2 align="center">
                Predictive performance of the PRS (genotype-only model) in hold-out test set
            </h2>
	        <p>
                The hold-out test set consists of unrelated individuals of white British ancestry in UK Biobank.
            </p>

            <a href="{{ snpnet_plot_pdf }}" target="_blank">
                <img class="img-responsive" src={{ snpnet_plot_png |tojson|safe }} alt={{ icd_str|tojson|safe }}>
            </a>
            <hr/>
        </div>
    </div>

    <div class="row">
        <div class="col col-md-12">
            <h2 align="center">Coefficients of the PRS model</h2>
            <br>
            <div id="containergenome2"></div>
            <br>
            Note: only <p id="number_of_active_variants" style="display:inline"></p> variants in active path for Lasso are included in the coefficient plot.
            <br>
            <hr/>
        </div>
    </div>

    <div class="row">
        <div class="col col-md-12">
            <h2 align="center">Predictive performance and transferability in UK Biobank</h2>
	        <p>
                The snpnet PRS model was developed using the training set (model development set)individuals of white British ancestry.
                The predictive performance was evaluated on a hold-out test set of white British ancestry. The transferability of the model was evaluated on the additioal sets of individuals of other ancestry groups in UK Biobank.
                For the definition of ancestry groups, we used the population stratification criteria described in our previous publication (<a href="https://rdcu.be/cdMmM" target="_blank">Sinnott-Armstrong*, Tanigawa*, et al. 2021</a>).
            </p>
            <br>
            <div align="center" id="container_table_eval" style="font-size:140%;">
                <table id="table_prs_trait_list">
                    <thead>
                        <tr>
                            <th rowspan="2">Metric</th>
                            <th rowspan="2">Ancestry group</th>
                            <th colspan="4">Predictive performance</th>
                            {% for table_col in table_eval_n_cols %}
                            <th rowspan="2">{{ table_col }}</th>
                            {% endfor %}
                        </tr>
                        <tr>
                            <th>Geno</th>
                            <th>Covars</th>
                            <th>Full</th>
                            <th>delta</th>
                        </tr>
                    </thead>
                    <tbody>
                        {{ table_eval_tbody_str|safe }}
                    </tbody>
                </table>
            </div>
        </div>
    </div>

    <div class="row">
        <div class="col col-md-12">
            <h3>Description of the columns</h3>
            <ul>
                <li>
                    Metric: the evaluation metric used in the predictive performance assessment. For quantitative traits, we used R2. For binary traits, we used ROC-AUC as well as pseudo-R2 metrics (Tjur's pseudo-R2 and Nagelkerke's pseudo-R2).
                </li>
		        <li>
                    Ancestry group: the set of individuals used for the evaluation in UK Biobank.
                </li>
                <li>
                    Geno: the predictive performence of the predictive model with genotype features alone. Note that this model does not include genotype PCs.
                </li>
                <li>
                    Covar: the predictive performence of the predictive model with covariates (age, sex, and genotype PCs) alone.
                </li>
                <li>
                    Full: the predictive performence of the full model that have both genotypes and covariates alone.
                </li>
                <li>
                    delta: the incremental predictive performence of the full model against the covariate-only model, i.e. the difference between the values in "Full" and "Covar" columns.
                </li>
                <li>
                    n: the number of individuals
                </li>
                <li>
                    case_n: (binary traits only) the number of cases
                </li>
                <li>
                    control_n: (binary traits only) the number of controls
                </li>
            </ul>
            <br>
	    <hr/>
	    <h2>References</h2>
	    <p>
		    Please check our preprint and paper for additional details of the PRS models, evaluation metrics, and the BASIL algorithm implemented in the snpnet package.
	    </p>
	    <p>
		    Y. Tanigawa, J. Qian, G. R. Venkataraman, J. M. Justesen, R. Li, R. Tibshirani, T. Hastie, M. A. Rivas, Significant Sparse Polygenic Risk Scores across 428 traits in UK Biobank. medRxiv 2021.09.02.21262942 (2021). <a href="https://doi.org/10.1101/2021.09.02.21262942" target="_blank">doi:10.1101/2021.09.02.21262942</a>
   	    </p>
        <p>
		    J. Qian, Y. Tanigawa, W. Du, M. Aguirre, C. Chang, R. Tibshirani, M. A. Rivas, T. Hastie, A fast and scalable framework for large-scale and ultrahigh-dimensional sparse regression with application to the UK Biobank. PLOS Genetics. 16, e1009141 (2020). <a href="https://doi.org/10.1371/journal.pgen.1009141" target="_blank">doi:10.1371/journal.pgen.1009141</a>
        </p>
        <hr/>
        </div>
    </div>
</div>
{% endblock %}

{% block endofbody %}
<script>
    var cutoff = 7;
    var activeset = 0;
    var activenum = 1;
    // manhattan plot code obtained/edited/adapted from : https://github.com/chengsoonong/rede/tree/master/Rede/manhattanplot
    d3.select("#containergenome2")
            .style("height", "100%")
            .style("overflow", "auto");

    var chrom_length = 0;
    var chrom_acum_length = new Array();
    var chromLength = new Array(
        249250621, 243199373, 198022430, 191154276,
        180915260, 171115067, 159138663, 146364022,
        141213431, 135534747, 135006516, 133851895,
        115169878, 107349540, 102531392, 90354753,
        81195210, 78077248, 59128983, 63025520,
        48129895, 51304566, 155270560, 59373566);

    //this initializes chrom_length and chrom_acum_length to be used in manhattan plot
    for (var i = 0; i < chromLength.length; i++) {
        chrom_length = chrom_length + chromLength[i];
        chrom_acum_length.push(chrom_length);
    }

    posarr = new Array();
    data_low = new Array();
    data_high = new Array();
    pvalarr = new Array();
    idarr = new Array();
    data_weight_pvalue = [];
    d3.tsv("{{ snpnet_BETAs_tsv }}",function(error,data){
        data.forEach(function(d) {
    // Fill two seperate arrays, one with points greater than the -log10(pvalue) threshold, and another
    // with points lower than the threshold defined in cutoff
            d.chrom = +d["CHROM"];
            d.pos = +d["POS"];
            d.beta = +d["BETA"];
            if (d.beta === 0) {
                return;
            }
            activeset = activeset + 1;
            d.id = d["CHROM"] + ":" + d["POS"] + "-" +  d["REF"] + "-" + d["ALT"];
            if (d.chrom == 1) {
                if (d.beta > -.01 && d.beta < .01) {
                    data_low.push( [d.pos, d.beta, d.id, d.chrom]);
                } else {
                    data_high.push([d.pos, d.beta, d.id, d.chrom]);
                }
            } else {
                if (d.beta > -.01 && d.beta < .01) {
                    data_low.push( [d.pos + chrom_acum_length[d.chrom - 2], d.beta, d.id, d.chrom]);
                } else {
                    data_high.push([d.pos + chrom_acum_length[d.chrom - 2],	d.beta, d.id, d.chrom]);
                }
            }
            data_weight_pvalue.push(d.beta);
        })

        // Merge the both high and low significance points together into a single array
        all_data = data_low.concat(data_high);
        var min_pvalue = d3.max([-3, d3.min([-.005, d3.min(data_weight_pvalue)])]);
        var max_pvalue = d3.max([.005, d3.max(data_weight_pvalue)]);
        // var which defined the extra space in the bottom and the top of the -log(p-value) axis dynamically to the dataset
        var extend_scale = (max_pvalue - min_pvalue) *0.05;
        ix_1 = 0;
        ix_2 = chrom_length;
        iy_1 = min_pvalue - extend_scale;
        iy_2 = max_pvalue + extend_scale;

        //create the manhattan plot
        x1 = ix_1;
        x2 = ix_2;
        y1 = iy_1;
        y2 = iy_2;

        var margin = {
            top: 30,
            right: 50,
            bottom: 20,
            left: 60
        };

        var colorScale = d3.scale.log()
            .domain([d3.min(all_data, function(d) {
                return parseInt(d[3]);
            }), d3.max(all_data, function(d) {
                return parseInt(d[3]);
            })])
            .interpolate(d3.interpolateHsl)
            .range(["#B3995D", "#8C1515"]);

        var w = 1100 - margin.left - margin.right; //900;
        var h = 500 - margin.top - margin.bottom; //600;

        //Create scale functions
        var xScale = d3.scale.linear()
            .domain([x1, x2])
            .range([0, w]);

        var yScale = d3.scale.linear()
            .domain([y1, y2])
            .range([h, 0]);

        var array_test1 = [""];
        var array_test2 = [0];

        for (var i = 0; i < chrom_acum_length.length; i++) {
            var num = i + 1;
            array_test1.push("chr" + num);
            array_test2.push(xScale(chrom_acum_length[i]));
        }
        //  console.log(x1,x2,y1,y2);

        //Create scale top
        var xScale_top = d3.scale.ordinal()
            .domain(array_test1)
            .range(array_test2);

        //Define X axis top
        var xAxis_top = d3.svg.axis()
            .scale(xScale_top)
            .orient("top")
            .ticks(0);

        //Define X axis
        var xAxis = d3.svg.axis()
            .scale(xScale)
            .orient("bottom");

        //Define Y axis
        var yAxis = d3.svg.axis()
            .scale(yScale)
            .orient("left")
            .ticks(5);

        // Define tooltip
        var tooltip = d3.select("#containergenome2")
            .append("div")
            .style("position", "fixed")
            .style("z-index", "0")
            .style("visibility", "hidden")
            .text("a simple tooltip");

        //Create SVG element
        var svg = d3.select("#containergenome2")
            .append("svg")
            .attr("width", w + margin.left + margin.right)
            .attr("height", h + margin.top + margin.bottom)
            .append("g")
            .attr("transform", "translate(" + margin.left + "," + margin.top + ")");

        // Vertical chromosome lines
        var line_chrom = svg.selectAll("line")
            .data(chrom_acum_length)
            .enter()
            .append("line")
            .attr("class", "linechrom")
            .attr("x1", function(d) {
                return xScale(d);
            })
            .attr("y1", 0)
            .attr("x2", function(d) {
                return xScale(d);
            })
            .attr("y2", h)
            .attr("stroke-width", 1)
            .attr("stroke-dasharray", 5)
            .style("stroke", "black") //stroke-dasharray="5"
            .style("opacity", 0.5);

        //Create circles for the low significance points
        var circle_low = svg.selectAll("circle_low")
            .data(data_low)
            .enter()
            .append("circle")
            .attr("cx", function(d) {
                return xScale(parseFloat(d[0]));
            })
            .attr("cy", function(d) {
                return yScale(parseFloat(d[1]));
            })
            .attr("r", 1)
            .style("fill", function(d) {
                return "#D3D3D3";
            });

        // Create circles for the high significance points
        var circle_high = svg.selectAll("circle_high")
            .data(data_high)
            .enter()
            .append("circle")
            .attr("cx", function(d) {
                return xScale(parseFloat(d[0]));
            })
            .attr("cy", function(d) {
                return yScale(parseFloat(d[1]));
            })
            .attr("r", function(d){
                return d3.max([2,1.5*parseFloat(d[1])]);
                })
            .on("mouseover", function(d){return tooltip.style("top", (event.pageY ) + "px").style("left", (event.pageX) + "px").style("visibility", "visible").text("variant id: " + d[2] + "; " + "Lasso coefficient: " + Number((d[1]).toFixed(4)));})
            .on("mousemove", function(){return tooltip.style("top", (event.pageY)+"px").style("left",(event.pageX)+"px");})
            .on("mouseout", function(){return tooltip.style("visibility", "hidden");})
            .on("click", function(d){
            window.open("https://biobankengine.stanford.edu/{{ namespace }}/awesome?query=" + d[2], '_blank');
            })
            .style("fill", function(d) {
                return colorScale(parseInt(d[3]));
            });

        //Create X axis
        svg.append("g")
            .attr("class", "manaxis").attr("font-size", "10px")
            .attr("transform", "translate(0," + (h) + ")")
            .call(xAxis);

        svg.append("g").attr("transform", "translate(0," + 0 + ")")
            .attr("class", "xt axis")
            .call(xAxis_top);

        svg.selectAll(".xt text") // select all the text elements for the xaxis
            .attr("transform", function(d) {
                return "translate(" + this.getBBox().height + "," + this.getBBox().height * -0.5 + ")rotate(-45)";
            });

        //Create Y axis
        svg.append("g")
            .attr("class", "manaxis")
            .call(yAxis)
            .append("text")
            .attr("class", "manlabel")
            .attr("transform", "rotate(-90)")
            .attr("x", -10)
            .attr("y", -50)
            .attr("dy", ".71em")
            .style("text-anchor", "end").attr("font-size", "17px")
            .text("Lasso coefficient");
        document.getElementById("number_of_active_variants").innerHTML = activeset.toString();
    });
</script>
{% endblock %}
